objective-c/README.md

```diff
@@ -0,0 +1,522 @@
+
+Example:
+
+	- (BOOL)openFile:(NSError * __autoreleasing *)error;	// YES, if the file is opened successfully (and error is nil) or NO as return value plus an NSError instance as error.
+
+How to implement?
+
+Example:
+
+	- (BOOL)openFile:(NSError * __autoreleasing *)error
+	{
+		BOOL success = NO;	// Will be the return value
+		NSError * internalError;
+
+		... // success flag & internal error can be set in here
+		
+		// Assign error value
+		if (error) {	// We need to check it because it can be nil
+			*error = [internalError copy];	// Watch out: *error = internalError; is not good enough!
+		}
+
+		return success;
+	}
+
+> Avoid to throw exception! The debugger will pause the running even you have a _@try_ around that point.
+
+## Metadata vs model objects
+
+Create objects from metadata as soon as possible!
+
+Example: Getting response JSON from the server:
+
+	[
+		{
+			id : 4
+			title : "How to develop application for iPhone"
+			pages : 358
+		}, 
+		...
+	]
+
+After the serialization we get an _NSArray_ with _NSDictionaries_.
+
+Of course we use named constants for the keys like:
+
+	static NSString * const MYDownloaderServerResponseKeyBookIdKey = @"id";
+	static NSString * const MYDownloaderServerResponseKeyBookTitleKey = @"title";
+	...
+
+We could make these public and use the dictionary everywhere, but we have more clean code if use 
+model object:
+
+MYBook.h
+
+	@interface MYBook : NSObject
+
+	@property (nonatomic, readonly, strong) NSNumber *bookId;
+	@property (nonatomic, readonly, strong) NSString *title;
+	@property (nonatomic, readonly, strong) NSNumber *pages;
+
+	+ (instancetype)bookWithId:(NSNumber *)bookId title:(NSString *)title pages:(NSNumber *)pages;
+	// and / or 
+	- (instancetype)initBookWithId:(NSNumber *)bookId title:(NSString *)title pages:(NSNumber *)pages;
+
+	@end
+
+## Macros < C-functions
+
+[Macros are dangerous](http://stackoverflow.com/questions/14041453/why-are-preprocessor-macros-evil-and-what-are-the-alternatives) but those easily can be replaced with C functions.
+
+	#define ADD(a, b) ...
+
+Replace with:
+
+	static NSInteger ADD(NSInteger a, NSInteger b) 
+	{ ... }
+
+
+
+## [Lambda expressions - alias "blocks"](https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/Blocks/index.html)
+
+Use blocks instead of delegate callbacks for singleton classes.
+
+> Avoid to retain self in block!
+
+	__weak MYViewController *weakSelf = self;
+	[[MYDownloader sharedInstance] start:^(NSArray<MYObject *> *result, NSError error) {
+		if (error) {
+			[weakSelf showErrorMessage:error];
+		} else {
+			[weakSelf updateContent:result];
+		}
+	}];
+
+## Some useful UIKit patterns
+
+### Properties of UIViewController
+
+Let's see this example:
+
+MYViewController.h
+
+	@interface MYViewController : UIViewController
+
+	@property (nonatomic, readonly, strong) UIView *backgroundView;
+
+	@end
+
+MYViewController.m
+
+	@implementation MYViewController
+
+	@syntesize backgroundView = _backgroundView;
```

Comment:
Well, depends on what we want to achieve with that read-only property.

If I needed a method to *create* a backgroundView I'd use a method like

    - (UIView *)buildBackgroundView {...}

This creates a new instance of UIView every time I send this message.

But in this case I want to get the `backgroundView` (which was already created) every time I ask for `self.backgroundView`, so I need to store it somewhere - and also need the `@syntesize myProperty;` because the [compiler doesn't do it for me in this case](https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/EncapsulatingData/EncapsulatingData.html#//apple_ref/doc/uid/TP40011210-CH5-SW13).